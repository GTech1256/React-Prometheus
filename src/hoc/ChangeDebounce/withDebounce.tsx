import React, { useCallback, useState, useEffect } from 'react'


function useDebounce(value: any, delay = 300) {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState<any>(value);

  useEffect(
    () => {
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, delay);

      return () => {
        clearTimeout(handler);
      };
    },
    [delay, value] 
  );

  return debouncedValue;
}

/**
 * Обертка над onChange каллбэком
 * 
 * Без onChange бессмысленно использовать этот HOC
 * */
// TOOD: Нормальные дженерики написать без <V extends unkown>
const withDebounce = (Component: React.ComponentType<any>, delay = 500) =>
  class extends React.Component<any, any> {
    debouncedValue = undefined
    currentTimeout?: NodeJS.Timeout = undefined

    handleChange = (value: any) => {
      if (this.currentTimeout) {
        clearTimeout(this.currentTimeout);
      }

      this.currentTimeout = setTimeout(() => {
        this.props.onChange(value);
      }, delay);
    }

    render() {
      return <Component {...this.props} onChange={this.handleChange} />
    }
  }

export default withDebounce