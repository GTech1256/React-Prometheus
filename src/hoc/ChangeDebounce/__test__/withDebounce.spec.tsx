import React from 'react';
import renderer from 'react-test-renderer';

import withDebounce from '../withDebounce'

const SomeComponent = () => <input />

describe('HOC: withDebounce', () => {
  it('should render correct', () => {
    const SomeComponentWithDebounce = withDebounce(SomeComponent);

    const tree = renderer
      .create(<SomeComponentWithDebounce />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  })
})