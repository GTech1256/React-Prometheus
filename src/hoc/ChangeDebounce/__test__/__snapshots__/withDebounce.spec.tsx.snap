// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repositories /> should render correct 1`] = `<input />`;

exports[`HOC: withDebounce should render correct 1`] = `<input />`;
